File: Queues-and-Deques/RandomizedQueue.java
import edu.princeton.cs.algs4.StdOut;
import edu.princeton.cs.algs4.StdRandom;

import java.util.Iterator;
import java.util.NoSuchElementException;

public class RandomizedQueue<Item> implements Iterable<Item> {

    private Item[] arr;
    private int count;


    // construct an empty randomized queue
    public RandomizedQueue() {
        arr = (Item[]) new Object[2]; // Init new object w/ 2 entries
        count = 0;
    }

    // is the randomized queue empty?
    public boolean isEmpty() {
        return count == 0;
    }

    // return the number of items on the randomized queue
    public int size() {
        return count;
    }

    private void resize(int cap) {
        Item[] temp = (Item[]) new Object[cap]; // Init new object w/ argument capacity
        for (int i = 0; i < count; i++) {
            temp[i] = arr[i];
        }
        arr = temp;
    }

    // add the item
    public void enqueue(Item item) {
        if (item == null) throw new IllegalArgumentException();
        if (count == arr.length) {
            resize(arr.length * 2); // double the size of the array when full
        }
        arr[count++] = item;
    }

    // remove and return a random item
    public Item dequeue() {
        if (isEmpty()) throw new NoSuchElementException("Queue underflow");
        int rand = StdRandom.uniform(0, count);
        Item oldItem = arr[rand];
        arr[rand] = arr[count - 1];
        arr[count - 1] = null;
        count--;
        if (count > 2 && count == arr.length / 4) {
            resize(arr.length / 2); // half the size of the array when array is 25% full
        }
        return oldItem;
    }

    // return a random item (but do not remove it)
    public Item sample() {
        if (isEmpty()) throw new NoSuchElementException("Queue underflow");
        int rand = StdRandom.uniform(0, count);
        return arr[rand];
    }

    // return an iterator over items in order from front to back
    public Iterator<Item> iterator() {
        return new LinkedIterator();
    }

    private class LinkedIterator implements Iterator<Item> {
        private Item[] current; // copy of arr so that multiple iterators can exist
        private int j = 0; // idx tracker

        public LinkedIterator() {
            copy();
            StdRandom.shuffle(current);
        }

        private void copy() {
            current = (Item[]) new Object[count];
            for (int i = 0; i < count; i++) {
                current[i] = arr[i];
            }
        }

        public boolean hasNext() {
            return j < count;
        }

        public void remove() {
            throw new UnsupportedOperationException();
        }

        public Item next() {
            if (!hasNext()) throw new NoSuchElementException();
            return current[j++];
        }
    }

    // unit testing (required)
    public static void main(String[] args) {
        int n = 5;
        RandomizedQueue<Integer> rq = new RandomizedQueue<Integer>();
        rq.enqueue(57);
        StdOut.println(rq.size());
        rq.dequeue();
        StdOut.println(rq.isEmpty());
        StdOut.println(rq.isEmpty());
        rq.enqueue(289);
        rq.dequeue();
        StdOut.println(rq.isEmpty());
        rq.enqueue(342);


    }

}

File: Princeton-Algorithms-I/Queues-and-Deques/RandomizedQueue.cpp
#include "RandomizedQueue.h"

template <typename Item>
RandomizedQueue<Item>::RandomizedQueue() {
    elems = new Item[2]; // Init new object w/ 2 entries
    length = 2;
    count = 0;
}

template <typename Item>
RandomizedQueue<Item>::~RandomizedQueue(){}

// is the randomized queue empty?
template <typename Item>
bool RandomizedQueue<Item>::isEmpty() {
        return count == 0;
}

// return the number of items on the randomized queue
template <typename Item>
int RandomizedQueue<Item>::size() {
    return count;
}

// Dynamically resize the storage array using repeated doubling
template <typename Item>
void RandomizedQueue<Item>::resize(int cap) {
    Item* temp = new Item[cap]; // Init new object w/ argument capacity
    for (int i = 0; i < count; i++) {
        temp[i] = elems[i];
}
    elems = temp;
    length = cap;
}

// add the item
template <typename Item>
void RandomizedQueue<Item>::enqueue(Item item) {
    if (item == NULL) throw "Illegal Argument: NULL";
    if (count == length) {
        resize(length * 2); // double the size of the array when full
    }
    elems[count++] = item;
}

// remove and return a random item
template <typename Item>
Item RandomizedQueue<Item>::dequeue() {
    if (isEmpty()) 
        throw "Queue underflow";
    int random = rand() % count; // generate random number from 0:count
    Item oldItem = elems[random];
    elems[random] = elems[count - 1];
    elems[count - 1] = NULL;
    count--;
    if (count > 2 && count == length / 4) {
        resize(length / 2); // half the size of the array when array is 25% full
    }
    return oldItem;
}

    // return a random item (but do not remove it)
template <typename Item>
Item RandomizedQueue<Item>::sample() {
    if (isEmpty()) 
        throw "Queue underflow";
    int random = rand() % count;
    return elems[random];
}



int main(int argc, char ** argv){

    srand (time(NULL));
	RandomizedQueue<char> rq;

	std::cout << "Empty? " << rq.isEmpty() << std::endl;
	std::cout << "Size? " << rq.size() << std::endl;
	rq.enqueue('a');
	rq.enqueue('b');
	rq.enqueue('c');
	rq.enqueue('d');
	std::cout << "Empty? " << rq.isEmpty() << std::endl;
	std::cout << "Size? " << rq.size() << std::endl;
   	int i;

	std::cout << "sampling : " ;
	for(i = 0; i < rq.size()*2;i++){
		std::cout << rq.sample() << " ";
	}
	std::cout << std::endl;

	std::cout << "dequeue :  " << rq.dequeue() << std::endl;

	return 0;
}